File: render/vue/src/types/execution.ts
/**
 * 代码执行相关类型定义
 *
 * @file src/types/execution.ts
 * @author WaterRun
 * @date 2026-02-10
 */

import type { Language } from './api'

/**
 * 执行状态
 *
 * @typedef
 */
export type ExecutionStatus =
    | 'pending'
    | 'running'
    | 'success'
    | 'error'
    | 'timeout'

/** 状态对应的显示文案 */
export const EXECUTION_STATUS_LABELS: Record<ExecutionStatus, string> = {
    pending: '等待中',
    running: '运行中',
    success: '成功',
    error: '错误',
    timeout: '超时',
}

/**
 * 执行记录
 *
 * @interface
 */
export interface Execution {
    /** 记录 ID */
    id: number
    /** 关联的模板 ID（无模板时为 null） */
    templateId: number | null
    /** 执行语言 */
    language: Language
    /** 源代码 */
    code: string
    /** 执行状态 */
    status: ExecutionStatus
    /** 标准输出 */
    output: string | null
    /** 错误信息 */
    error: string | null
    /** 执行耗时（毫秒） */
    durationMs: number | null
    /** 创建时间（ISO 8601） */
    createdAt: string
}

/**
 * 提交执行请求
 *
 * @interface
 */
export interface ExecuteRequest {
    /** 源代码 */
    code: string
    /** 语言 */
    language: Language
}

/**
 * 提交执行响应
 *
 * @interface
 */
export interface ExecuteResponse {
    /** 记录 ID */
    id: number
    /** 执行状态 */
    status: ExecutionStatus
    /** 标准输出 */
    output: string | null
    /** 错误信息 */
    error: string | null
    /** 执行耗时（毫秒） */
    durationMs: number | null
}

File: server/release/src/model/execution.py
r"""
执行相关模型定义

:file: src/model/execution.py
:author: WaterRun
:time: 2026-03-15
"""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field


class ExecutionStatus(StrEnum):
    r"""
    执行状态枚举
    """

    SUCCESS = "Success"
    ERROR = "Error"
    TIMEOUT = "Timeout"


class SupportedLanguage(StrEnum):
    r"""
    支持的编程语言枚举
    """

    PYTHON = "python"
    LUA = "lua"
    RUST = "rust"
    C = "c"


class ExecuteRequest(BaseModel):
    r"""
    代码执行请求体
    """

    language: SupportedLanguage = Field(description="编程语言")
    code: str = Field(description="待执行的代码")


class ExecuteResponse(BaseModel):
    r"""
    代码执行响应
    """

    execution_id: int = Field(description="执行记录ID")
    status: ExecutionStatus = Field(description="执行状态")
    toml_output: str | None = Field(default=None, description="生成的TOML内容")
    error_message: str | None = Field(default=None, description="错误信息")
    execution_time: int = Field(description="执行时间，单位毫秒")
    cached: bool = Field(description="是否命中缓存")


class ExecutionHistoryItem(BaseModel):
    r"""
    执行历史列表项
    """

    id: int = Field(description="执行记录ID")
    language: SupportedLanguage = Field(description="编程语言")
    code: str = Field(description="源代码")
    status: ExecutionStatus = Field(description="执行状态")
    execution_time: int | None = Field(default=None, description="执行时间，单位毫秒")
    created_at: datetime = Field(description="执行时间")


class ExecutionDetail(BaseModel):
    r"""
    执行记录详情
    """

    id: int = Field(description="执行记录ID")
    language: SupportedLanguage = Field(description="编程语言")
    code: str = Field(description="源代码")
    toml_output: str | None = Field(default=None, description="生成的TOML内容")
    status: ExecutionStatus = Field(description="执行状态")
    error_message: str | None = Field(default=None, description="错误信息")
    execution_time: int | None = Field(default=None, description="执行时间，单位毫秒")
    created_at: datetime = Field(description="执行时间")


class ExecutionListResponse(BaseModel):
    r"""
    执行历史分页响应
    """

    items: list[ExecutionHistoryItem] = Field(description="执行历史列表")
    total: int = Field(description="总数")
    page: int = Field(description="当前页码")
    limit: int = Field(description="每页数量")
